### ISSUE TITLE
Investigate case where ancestor chain gets reordered

### ISSUE THREAD

**Issue by JacobEvelyn** [panorama-ed/memo_wise#135]
We encountered a case in a complex app where we saw something like the following:

```ruby
class ExampleClass
  include ExampleModule

  def example_method
    true
  end
  memo_wise :example_method
end

module ExampleModule
  prepend MemoWise
  extend ActiveSupport::Concern

  included do
    def example_method
      false
    end
  end
end

ExampleClass.new.example_method
# => true
```

And we noticed that if instead we have:

```ruby
class ExampleClass
  prepend MemoWise # This line added
  include ExampleModule

  ...
end

# Now:
ExampleClass.new.example_method
# => false
```

but if we move the `prepend MemoWise` after the `include ExampleModule` we get the original behavior again. Something about calling `prepend MemoWise` first modifies the ancestor chain in a way I do not understand.

We should investigate and see if this is expected Ruby behavior, a bug in MemoWise, or a bug in the Ruby interpreter.

**Comment by LeonardoGodoy** [comment]
hey @JacobEvelyn, which was the ruby version you used? 
I tried to simulate that with 2.6.3 and it returned true in both cases.

Either way, I simulated all combinations prepending and not prepending the Memowise module on the module example.
In the scenario where it prepended both in the ExampleClass and ExampleModule, and it is done in the class after the inclusion of ExampleModule the ancestors chain looks exactly like MemoWIse was **included** on ExampleClass.

I created a repo with the tests I did if you want to check it out https://github.com/LeonardoGodoy/memo_wise_issue_135

**Comment by LeonardoGodoy** [comment]
Interesting... I simulate that behavior with a non-memorise module and the ancestor's chan was not affected the same way.
Alter that I tested if it was active_support/concern and it was not either.
I will get back to it asap

**Comment by JacobEvelyn** [comment]
Hi @LeonardoGodoy! Thanks for digging in here and sorry for the slow response. This was in Ruby 2.6.6. I can spend some time this week digging into our (closed-source, sadly) codebase to see if I can come up with an actual reproducible test case.

**Comment by PikachuEXE** [comment]
I might have encountered the same issue
```
module TestModule1
  prepend ::MemoWise

  def pika?
    :chu
  end
  memo_wise :pika?
end


class TestClass1
  prepend ::MemoWise
  include TestModule1

  private def pika?
    :pikapika
  end
end

TestClass1.new.send(:pika?) #=> unexpected `:chu` instead of `:pikapika`

TestClass1.ancestors
# Actual: [MemoWise, TestModule1, TestClass1, Object, PP::ObjectMixin, Kernel, BasicObject]
# Expected: [MemoWise, TestClass1, TestModule1, Object, PP::ObjectMixin, Kernel, BasicObject]
```

**Comment by PikachuEXE** [comment]
This issue seems fixed in ruby-master (using [these instructions](https://github.com/ruby/ruby/blob/master/doc/contributing.rdoc#running-tests-) to build)
I will test more with different branches and update this comment
![image](https://user-images.githubusercontent.com/1018543/128130031-075014e0-5fe0-4901-879d-7fd15e42a468.png)

**Comment by JacobEvelyn** [comment]
Thanks for the investigation, @PikachuEXE ! I hadn't finished investigating/reproducing this within our codebase but yes this does seem like the same issue. I'm going to close this issue as it doesn't appear to be related to `MemoWise`. Feel free to comment if you disagree!

**Comment by PikachuEXE** [comment]
Happy to see https://github.com/ruby/ruby/pull/4711 merged and released in `2.7.5`
I have run the test script in `irb` and confirmed the issue no longer exists in `2.7.5`

**Comment by JacobEvelyn** [comment]
Awesome! Thanks for the update @PikachuEXE and glad to hear the backport finally merged!
